### PR TITLE
Add BoltLockManager and Silabs Cluster implementation 

### PR DESCRIPTION
**Pull request by jmartinez-silabs**
#### Problem
Add BoltLockManager and Silabs Cluster implementation 

 #### Summary of Changes
Add support for Silabs dev board BRD4163A and BRD4164A 
Add BoltLockManager to manage the lock and unlock request and state
Add DataModelHandler to handle bolt actions from the cluster messages
Add Gen folder with the files for silicon lab cluster implementation
Start a server session for UDP messages
Include some mbedtls sources  files from gsdk 2.7 in gni . TO BE FIX

### CODE REVIEW

**Comment by rwalker-apple** [comment]
this is kinda neat, mapping on-off to a lock.

the "gen" directory is copied from another example, or is generated and modified a la wifi echo server?

**Review by rwalker-apple — APPROVED** [review]
lgtm, but you might want to file an issue that this should be using the lock cluster at some point

**Comment by jmartinez-silabs** [comment]
@rwalker-apple The gen directory has been copied from the work on same example from NRF52. I will create a issue to do the work to map the lock cluster instead. We are working on having the example work with openthread by the end of the week

By the way, I did a miss step with pull and rebase that caused undesired change on submodules repo. That is why some test were failing. I fixed them by checking out the current commit version of those repo to revert that change. Is this correct. I don't want to mess with the submodules

**Review by mspang — APPROVED** [review]

**Comment by jmartinez-silabs** [comment]
@woody-apple @robszewczyk @andy31415 @bzbarsky-apple. Could you guys review this we want to complet the example by the end of the week and we have others PR comming to add to this one. Thank you

**Review by woody-apple — APPROVED** [review]

**Comment by woody-apple** [comment]
@saurabhst @jelderton @andy31415 ?

**Review by jelderton — APPROVED** [review]

File: examples/lock-app/efr32/include/DataModelHandler.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+ * @file
+ *   This file defines the API for the handler for data model messages.
+ */
+
+#ifndef DATA_MODEL_HANDLER_H
+#define DATA_MODEL_HANDLER_H
+
+namespace chip {
+namespace System {
+class PacketBuffer;
```
Comment (andy31415) [inline examples/lock-app/efr32/include/DataModelHandler.h:28]: Do we need this forward declaration in this header?

**Review by andy31415 — APPROVED** [review]

File: examples/lock-app/efr32/src/BoltLockManager.cpp [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+
+void BoltLockManager::SetCallbacks(Callback_fn_initiated aActionInitiated_CB, Callback_fn_completed aActionCompleted_CB)
+{
+    mActionInitiated_CB = aActionInitiated_CB;
+    mActionCompleted_CB = aActionCompleted_CB;
+}
+
+bool BoltLockManager::IsActionInProgress()
+{
+    return (mState == kState_LockingInitiated || mState == kState_UnlockingInitiated) ? true : false;
```
Comment (andy31415) [inline examples/lock-app/efr32/src/BoltLockManager.cpp:62]: nit: pattern of `foo ? true : false` can generally be replaced by just `foo` and result in shorter code, without loss of readability. Here and below.

File: examples/lock-app/efr32/src/Server.cpp [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+using namespace ::chip;
+using namespace ::chip::Inet;
+using namespace ::chip::Transport;
+
+// Transport Callbacks
+namespace {
+
+#ifndef EXAMPLE_SERVER_NODEID
+// "nRF5"
```
Comment (andy31415) [inline examples/lock-app/efr32/src/Server.cpp:54]: Is this comment supposed to be here? It does not help me understand the code better. Maybe needs removal or expansion.

**Review by saurabhst — APPROVED** [review]

File: examples/lock-app/efr32/src/BoltLockManager.cpp (hunk 2 above)
Comment (jmartinez-silabs) [reply examples/lock-app/efr32/src/BoltLockManager.cpp]: Updated with a clean up as mentionned

File: examples/lock-app/efr32/src/Server.cpp (hunk 3 above)
Comment (jmartinez-silabs) [reply examples/lock-app/efr32/src/Server.cpp]: Added a comment as we use this hardcoded nodeID for the example right now

File: examples/lock-app/efr32/include/DataModelHandler.h (hunk 1 above)
Comment (jmartinez-silabs) [reply examples/lock-app/efr32/include/DataModelHandler.h:28]: We have this header on pretty much all the files in the example. If not needed we could do a clean up in future PR

File: examples/lock-app/efr32/src/Server.cpp [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+     * @param [in] buffer The buffer holding the message.  This function guarantees
+     *                    that it will free the buffer before returning.
+     */
+    void HandleDataModelMessage(const MessageHeader & header, System::PacketBuffer * buffer, SecureSessionMgrBase * mgr)
+    {
+        EmberApsFrame frame;
+        bool ret = extractApsFrame(buffer->Start(), buffer->DataLength(), &frame);
+        if (ret)
+        {
+            EFR32_LOG("APS framprocessing success!");
```
Comment (bzbarsky-apple) [inline examples/lock-app/efr32/src/Server.cpp:126]: "framprocessing"?

**Review by bzbarsky-apple — COMMENTED** [review]
lgtm modulo the one nit

**Pull request merged**